### ISSUE TITLE
Discriminated lists do not work in v2

### ISSUE THREAD

**Issue by miili** [pydantic/pydantic#6453]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

Discriminated lists cannot do not work in v2

In v1 an annotated union can be used to discriminate elements in a list

In v2 this raises

```
TypeError: `discriminator` can only be used with `Union` type with more than one variant
```

I cannot find a way to implement this following https://docs.pydantic.dev/2.0/usage/errors/#discriminator-no-field

What is missing for this feature (which is really great)?

### Example Code

```Python
Features = Annotated[
    Union[GreatFeature, AwesomeFeature, UltraFeature],
    Field(..., discriminator="feature"),
]

class Model(BaseModel):
    feature_list: list[Features] = []
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0
        pydantic-core version: 2.0.1 release build profile
               python version: 3.11.2 (main, Apr 19 2023, 11:03:54) [GCC 10.2.1 20210110]
                     platform: Linux-5.10.0-23-amd64-x86_64-with-glibc2.31
     optional deps. installed: ['typing-extensions']
```

Selected Assignee: @hramezani

**Comment by Kludex** [comment]
I can't reproduce the code without the right imports. 🤷‍♂️ 

Can you try to bump pydantic to `2.0.1` instead of `2.0`? It's likely that we fixed this already.

**Comment by miili** [comment]
Am already on `2.0.1`.

debugging shows that the union identifies itself as `tagged-union`

```
{'type': 'tagged-union', 'choices': {...}}
```

This is from https://github.com/pydantic/pydantic/blob/main/pydantic/_internal/_discriminated_union.py#L183

Should this be
```py
schema['type'] not in ('union', 'tagged-union'):
    ...
```

I can provide a fix.

**Comment by Kludex** [comment]
> Am already on 2.0.1.

Hmmm... Not according to your issue description. It says `pydantic-core==2.0.1`, and `pydantic==2.0`. Is that wrong?

If you can provide the a MRE I can try it here. Also, this was opened some hours ago: https://github.com/pydantic/pydantic/issues/6444.

**Comment by miili** [comment]
dang, looked at `pydantic-core`... fixed in 2.0.1. Thanks!

**Comment by Kludex** [comment]
Thanks for coming back, and closing it! 🙏
